src/lava/magma/core/model/py/ports.py

```diff
@@ -208,9 +251,102 @@ def write(self, data: int, idx: int):
 PyRefPort.SCALAR_SPARSE = PyRefPortScalarSparse
 
 
-# ToDo: Remove... not needed anymore
-class _PyRefPort(Enum):
-    VEC_DENSE = PyRefPortVectorDense
-    VEC_SPARSE = PyRefPortVectorSparse
-    SCALAR_DENSE = PyRefPortScalarDense
-    SCALAR_SPARSE = PyRefPortScalarSparse
+class PyVarPort(AbstractPyPort):
+    """Python implementation of InPort used within AbstractPyProcessModel.
+    If buffer is empty, recv() will be blocking.
+    """
+
+    def __init__(self,
+                 var_name: str,
+                 csp_send_port: ty.Optional[CspSendPort],
+                 csp_recv_port: ty.Optional[CspRecvPort], *args):
+        self._csp_recv_port = csp_recv_port
+        self._csp_send_port = csp_send_port
+        self.var_name = var_name
+        super().__init__(*args)
+
+    VEC_DENSE: ty.Type["PyVarPortVectorDense"] = None
+    VEC_SPARSE: ty.Type["PyVarPortVectorSparse"] = None
+    SCALAR_DENSE: ty.Type["PyVarPortScalarDense"] = None
+    SCALAR_SPARSE: ty.Type["PyVarPortScalarSparse"] = None
+
+    def start(self):
+        if self._csp_send_port is not None:
+            self._csp_send_port.start()
+        if self._csp_recv_port is not None:
+            self._csp_recv_port.start()
+
+    def join(self):
+        if self._csp_send_port is not None:
+            self._csp_send_port.join()
+        if self._csp_recv_port is not None:
+            self._csp_recv_port.join()
+
+    def service(self):
+        pass
+
+
+class PyVarPortVectorDense(PyVarPort):
+    def service(self):
+        """Sets the received value to the given var or sends the value of the
+        var to the csp_send_port, depending on the received header information
+        of the csp_recv_port."""
+
+        # Inspect incoming data
+        if self._csp_send_port is not None and self._csp_recv_port is not None:
```

Comment:
You mean to delete an unconnected VarPort from the list of VarPorts? The list is created earlier then the csp channel creation. We already check for unconnected Ports and do not created attributes for those in the process model.

It is also checked that the list includes all attributes of type LavaPyType of a process model. Which would fail if we exclude based on connections.

Reply:
I guess this only really matters for explicit VarPorts... Implicit VarPorts are only ever created if the are actually needed and connected.
For implicit VarPorts, the VarPorts at the Process level can stay but we don't have to create or put the PyVarPorts in the ProcessModel in the list of PyVarPorts you must otherwise check each iteration.